src/AggregateFunctions/AggregateFunctionGroupSortedArray.h

```diff
@@ -0,0 +1,248 @@
+#pragma once
+
+#include <Columns/ColumnArray.h>
+#include <DataTypes/DataTypeArray.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <base/logger_useful.h>
+#include "AggregateFunctionGroupSortedArrayData.h"
+
+namespace DB
+{
+template <typename TT, bool is_plain_column>
+inline TT readItem(const IColumn * column, Arena * arena, size_t row)
+{
+    if constexpr (std::is_same_v<TT, StringRef>)
+    {
+        if constexpr (is_plain_column)
+        {
+            StringRef str = column->getDataAt(row);
+            auto ptr = arena->alloc(str.size);
+            std::copy(str.data, str.data + str.size, ptr);
+            return StringRef(ptr, str.size);
+        }
+        else
+        {
+            const char * begin = nullptr;
+            return column->serializeValueIntoArena(row, *arena, begin);
+        }
+    }
+    else
+    {
+        return column->getUInt(row);
+    }
+}
+/*
+template <typename T> struct ItemNTH
+{
+    uint64_t a;
+    T b;
+    bool operator < (const ItemNTH& other) const    {return (this->a < other.a);}
+};
+
+template <typename T> void getFirstNElements(T *data, size_t num_elements, size_t threshold, uint64_t *results)
+{
+    threshold = std::min(threshold, num_elements);
+
+    ItemNTH<T> *dataIndexed = new ItemNTH<T>[num_elements];
+    for (size_t i = 0; i < num_elements; i++) {
+        dataIndexed[i].a = data[i];
+        dataIndexed[i].b = i;
+    }
+
+    std::nth_element(dataIndexed, dataIndexed + threshold, dataIndexed + num_elements);
+    std::sort(dataIndexed, dataIndexed + threshold);
+
+    for (size_t i = 0; i < threshold; i++) {
+        results[i] = dataIndexed[i].b;
+    }
+
+    delete []dataIndexed;
+}
+*/
+
+template <typename T>
+void getFirstNElements(const T * data, int num_elements, int threshold, uint64_t * results)
+{
+    threshold = std::min(num_elements, threshold);
+    int current_max = 0;
+    int cur;
+    int z;
+    for (int i = 0; i < num_elements; i++)
+    {
+        //We starting from the highest values and we look for the immediately lower than the given one
+        for (cur = current_max; cur > 0 && (data[i] < data[results[cur - 1]]); cur--)
+            ;
+
+        if (cur < threshold)
+        {
+            //Move all the higher values 1 position to the right
+            for (z = current_max - 1; z >= cur; z--)
+                results[z + 1] = results[z];
+            if (++current_max > threshold)
```

Comment:
Wouldn't it be simpler, and avoid this check, if you prepopulate the results array with the first `threshold` elements, sort them, and then continue knowing that current_max is always == threshold?